File: csharp/src/Vista.SDK/Transport/TimeSeriesData/DataChannelId.cs
namespace Vista.SDK.Transport;

public readonly record struct DataChannelId
{
    private readonly int _tag;
    private readonly LocalId? _localId;
    private readonly string? _shortId;

    public readonly bool IsLocalId => _tag == 1;
    public readonly bool IsShortId => _tag == 2;

    public readonly LocalId? LocalId => _tag == 1 ? _localId : null;

    public readonly string? ShortId => _tag == 2 ? _shortId : null;

    private DataChannelId(LocalId value)
    {
        _tag = 1;
        _localId = value;
        _shortId = null;
    }

    private DataChannelId(string value)
    {
        _tag = 2;
        _localId = null;
        _shortId = value;
    }

    public readonly T Match<T>(Func<LocalId, T> onLocalId, Func<string, T> onShortId)
    {
        return _tag switch
        {
            1 => onLocalId(_localId!),
            2 => onShortId(_shortId!),
            _ => throw new InvalidOperationException("Tried to match on invalid DataChannelId"),
        };
    }

    public readonly void Switch(Action<LocalId> onLocalId, Action<string> onShortId)
    {
        if (_tag == 1)
            onLocalId(_localId!);
        else if (_tag == 2)
            onShortId(_shortId!);
        else
            throw new InvalidOperationException("Tried to switch on invalid DataChannelId");
    }

    public readonly override string ToString() =>
        _tag switch
        {
            1 => _localId!.ToString(),
            2 => _shortId!,
            _ => throw new Exception("Invalid state exception"),
        };

    public static DataChannelId Parse(string value)
    {
        if (value is null)
            throw new ArgumentNullException(nameof(value));

        if (LocalIdBuilder.TryParse(value, out var localIdBuilder))
            return new DataChannelId(localIdBuilder.Build());
        else
            return new DataChannelId(value);
    }

    public static implicit operator DataChannelId(LocalId id) => new DataChannelId(id);
}

File: js/src/vista-sdk/lib/transport/domain/time-series-data/DataChannelId.ts
import {
    Codebooks,
    Gmod,
    LocalId,
    LocalIdBuilder,
    LocalIdParsingErrorBuilder,
} from "../../..";
import { isNullOrWhiteSpace } from "../../../util/util";

export class DataChannelId {
    private readonly _tag: number;
    private readonly _localId?: LocalId;
    private readonly _shortId?: string;

    private constructor(value: LocalId | string) {
        if (typeof value === "string") {
            this._tag = 2;
            this._shortId = value;
        } else {
            this._tag = 1;
            this._localId = value;
        }
    }

    public get isLocalId(): boolean {
        return this._tag == 1;
    }

    public get isShortId(): boolean {
        return this._tag == 2;
    }

    public get localId() {
        return this._tag === 1 ? this._localId : undefined;
    }

    public get shortId() {
        return this._tag === 2 ? this._shortId : undefined;
    }

    public match<T>(
        onLocalId: (localId: LocalId) => T,
        onShortId: (shortId: string) => T
    ): T {
        switch (this._tag) {
            case 1:
                return onLocalId(this._localId!);
            case 2:
                return onShortId(this._shortId!);
            default:
                throw new Error("Tried to match on invalid DataChannelId");
        }
    }

    public switch(
        onLocalId: (localId: LocalId) => void,
        onShortId: (shortId: string) => void
    ): void {
        switch (this._tag) {
            case 1:
                onLocalId(this._localId!);
                break;
            case 2:
                onShortId(this._shortId!);
                break;
            default:
                throw new Error("Tried to switch on invalid DataChannelId");
        }
    }

    public toString(): string {
        switch (this._tag) {
            case 1:
                return this._localId!.toString();
            case 2:
                return this._shortId!;
            default:
                throw new Error("Invalid DataChannelId");
        }
    }

    public static parse(
        value: string | undefined,
        gmod: Gmod,
        codebooks: Codebooks,
        errorBuilder?: LocalIdParsingErrorBuilder
    ): DataChannelId {
        if (isNullOrWhiteSpace(value))
            throw new Error(`${this.parse.name}: value is null`);

        let localId: LocalIdBuilder | undefined = undefined;

        if (
            (localId = LocalIdBuilder.tryParse(
                value,
                gmod,
                codebooks,
                errorBuilder
            ))
        ) {
            return new DataChannelId(localId.build());
        } else {
            return new DataChannelId(value);
        }
    }

    public static async parseAsync(
        value: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        if (isNullOrWhiteSpace(value))
            throw new Error(`${this.parse.name}: value is null`);

        let localId: LocalIdBuilder | undefined = undefined;

        if (
            (localId = await LocalIdBuilder.tryParseAsync(value, errorBuilder))
        ) {
            return new DataChannelId(localId.build());
        } else {
            return new DataChannelId(value);
        }
    }
}
